### PR TITLE
Add issue table view

### LINKED ISSUE #423: Add issues table component to issues details page
## Context:

- The current implementation of the dataset issues page is that we're playing issues back to users on a per-entity / per-record basis, meaning that users have to click through each entity/record to view the details of the issue.
- This isn't a great user experience when the user is faced with a high volume of issues.
- A better experience would be to display this data back to them in an aggregated format, in a table.
- The intention is to re-use the component from the check tool, so that we don't have multiple versions of the same thing floating around, and only have to maintain styles in one place: https://github.com/digital-land/submit/issues/421

## UX:

- The table view will be the default view for the issues details page.
- Users will navigate to the issues entity page from the issues table. The table should include a link to the entity-level view, to each specific entity.
- The table view the same as the current entity-level view, except that the entity table is replaced with the aggregated table.
- Each view should have their own pagination as well (might not always need it shown if all rows fit on a single page).
- In terms of column ordering in the table: reference should show first, and ideally the rest should be in alphabetical order (though we may want to cut this from the scope depending on the level of complexity).
- To navigate back to the table view from the entity-level view, we will use a back link above the summary (see: https://github.com/digital-land/submit/issues/406#issuecomment-2344033857).

## Acceptance criteria:

- As a data provider, when I'm on the dataset issues page, I see a table that shows me an aggregated view of all the records with the issue that I'm viewing.
- As a data provider, when I'm on the dataset issues page, the table contains a link to each individual entity that takes me to the entity-level view.
- As a data provider, when I'm on the entity-level view of the dataset issues page, I see a link to take me back to the table view.

## Tasks:

- [x] write middleware's
- [x] make a basic template 
- [x] Populate error summary
- [x] populate breadcrumb links
- [x] have the reference's link through to the entity views
- [x] ~sort the columns to spec order~ (rendering in specification order)
- [x] duplicate reference issues should skip to entry view
- [x] don't show issues where there's no entity (except those to do with reference)
- [x] add pagination
- [x] add valibot schema
- [x] move test code for moved middlewares
- [ ] ~add test code for new template issueTable.html~ (going to pick this up in a new ticket as its complicated to mock the table params, and the dataview template also needs testing)
- [x] add code to test middleware in dataview.middleware.js
  - [x] getDataRange
- [x] add test code for middleware in common.middleware.js
  - [x] createPaginationTemplateParams
  - [x] show404IfPageNumberNotInRange
  - [x] FilterOutIssuesToMostRecent
  - [x] removeIssuesThatHaveBeenFixed
  - [x] addFieldMappingsToIssue
- [x] add test code for middleware in issueTable.middleware.js
  - [x] getDataRange
  - [x] prepareTableParams
  - [x] getErrorSummaryItems
  - [x] prepareTemplateParams
  - [x] notIssueHasEntity
  - [x] issueTypeAndFieldShouldRedirect
  - [x] redirectToEntityView
  - [x] getIssueTable
  - [x] show404IfPageNotInRange

# Content review tasks
- [x] On the pages that go to the record view, the summary box is empty https://submit-pr-691.herokuapp.com/organisations/local-authority:LBH/tree/missing%20value/point/entry
- [ ] table view summary showing column level error while task list is not
  - this is correct on the table view, but incorrect on the tasklist. will be resolved when updating the task list to look at all active resources/entities 
- [x] article 4 lambeth showing data on table but not on entity
  - this was due to a bug, have fixed in this PR
  - this page is also only looking at the latest active resource, ideally this should look at the entitiy similar to how the table view works
    - Alex has a new design for this page, so might just hold of to do it in there

## Problems:

> 
> ### duplicate entity references provided by different organisations
> data management are looking at this one
> 
> --------------------------------------------------------------
> 
> ### Any issue causing an entry not to become an entity
> we can't really show it on the table as it doesn't have any data, only the issues!
> I think we want a different view for this, I suggest we skip the table view and go directly to an individual task view
> Also, we should only pull these issues from the latest resource, as we don't know if they have been fixed in a later resource
> for example: https://submit.planning.data.gov.uk/organisations/local-authority:LBH/listed-building-outline
> 
> Now: skip to entry view
> Later: updated design indicating the resource and line number where this is happening
> 
> --------------------------------------------------------------
> 
> ### Duplicate entity references provided by the same organisation
> for this type of task, instead of having a page with for example three 'duplicate reference issues' we go to a page that has one issue stating something along the lines of: 'you have 3 entries all with the same reference across lines and 4, 20, 24'
> for examlpe here: https://submit-pr-691.herokuapp.com/organisations/local-authority:LBH/listed-building-outline
> 
> Now: skip to entry view
> Later: new design for each duplicate, showing the processed entity and indicating the resource and rows of the duplicates
> 
> --------------------------------------------------------------
> 
> ### Any other issue not on an entry that isn't an entity not causing the entity to be made
> I don't think we should be showing right now, its really complicated especially when some some of these issue types and field pairs are on an entity and some are not, after they fix the reference issue. then any other issues will show up
> http://localhost:5000/organisations/local-authority:LBH/article-4-direction-area
> This one becomes even more problematic when a previously uploaded entity without a reference has other errors and then the org fixes them by reuploading the file but deleting the rows. 
> for example, [lambeth has no article-4-direction areas](https://www.planning.data.gov.uk/entity/?dataset=article-4-direction-area&entry_date_day=&entry_date_month=&organisation_entity=192) but they previously tried to upload some, but didn't provide references. so we ended up with [these issues](http://localhost:5000/organisations/local-authority:LBH/article-4-direction-area) that can never be resolved
> 
> Now: don't show these issues
> 
> --------------------------------------------------------------
> 
> ### What do we do when an issue doesn't exist in an earlier resource, but it does in a later resource??
> So looking at [this](https://submit-pr-691.herokuapp.com/organisations/local-authority:MAV/brownfield-land/missing%20value/SiteNameAddress) we can see missing-value issues on fields that have values. this is because in a previous resource the values were there, but in their latest resource they have removed that value for some reason???
> looking at the [data](https://datasette.planning.data.gov.uk/digital-land?sql=select+*+from+resource+r%0D%0ALEFT+JOIN+resource_organisation+ro+ON+r.resource+%3D+ro.resource%0D%0ALEFT+JOIN+resource_dataset+rd+ON+r.resource+%3D+rd.resource%0D%0AWHERE+rd.dataset+%3D+%27brownfield-land%27%0D%0AAND+ro.organisation+%3D+%27local-authority%3AMAV%27%0D%0AAND+r.end_date+%3D+%27%27%0D%0AORDER+BY+r.start_date) basically, in row 3 the values are present but row 4 they are not 
> 
> Now: show issue for now with normal message
> Later: update messages to indicate this is occurring in your latest resource
> 
> --------------------------------------------------------------
> 
> ### The task list and overview is currently gathering issues that have been resolved and are no longer relevant
> this can be problematic as in a few edge cases issues that has been fixed by a newer resource are still being used i.e. https://submit.planning.data.gov.uk/organisations/local-authority:MAV/brownfield-land
> 
> Now: Deploy as normal, then start work on other pages to bring into sync with this one
>  
> 
> 
>

### PR DESCRIPTION
**Pull request by GeorgeGoodall-GovUk**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adds the issue table page
moved some code from issue details and dataview to common and refactored for reuse

## Related Tickets & Documents
- Closes #423 

## QA Instructions, Screenshots, Recordings
Preview Link: https://submit-pr-691.herokuapp.com/organisations/local-authority:WRT/brownfield-land/invalid%20URI/SiteplanURL/entry

image

## Added/updated tests?
- [x] Yes

## QA sign off
- [x] Code has been checked and approved
- [ ] Design has been checked and approved
- [ ] Product and business logic has been checked and proved

## [optional] Are there any post-deployment tasks we need to perform?
- We need to update other areas of the site to get issues using the new entity issue link. 
  - for now we have agreed to go ahead with this deployment, but pick up this other work as a priority

## Release Notes

- **New Features**
	- Introduced middleware for managing issue tables, including validation and data preparation.
	- Enhanced pagination handling with new functions for calculating data ranges and setting base paths.
	- Updated templates for displaying error summaries and issue details, improving user experience.
	- Added new schemas for structured data representation in requests and responses.
	- Implemented new middleware functions for streamlined processing of dataset specifications and entity issues.
	- Added a new property for pagination settings in the configuration.

- **Bug Fixes**
	- Improved error handling in templates for issue details and summaries.
	- Restructured pagination logic to ensure correct display based on available data.

- **Documentation**
	- Updated schemas for better data structure representation in requests and responses.

- **Tests**
	- Expanded test coverage for middleware functions and templates, ensuring robust validation of new features and bug fixes.

### CODE REVIEW

File: src/views/organisations/issueTable.html [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      <li>Publish the updated data on the data URL</li>
+    </ol>
+  </div>
+</div>
+{% endblock %}
+
+{% block scripts %}
+  {{ super() }}
+  {% if geometries and geometries.length %}
+    <link href='https://unpkg.com/maplibre-gl@latest/dist/maplibre-gl.css' rel='stylesheet' />
```
[the remark this answers is not in the document]
Comment (GeorgeGoodall-GovUk) [reply src/views/organisations/issueTable.html:112]: will leave for now, but ideally we should have this scc in our service. ticket made
https://github.com/orgs/digital-land/projects/7/views/11?pane=issue&itemId=89041757&issue=digital-land%7Csubmit%7C702

File: test/unit/views/organisations/issueDetailsPage.test.js [hunk 2]
```diff
@@ -177,8 +176,10 @@ describe(`issueDetails.html(seed: ${seed})`, () => {
       const paramWithGeometry = {
         organisation: params.organisation,
         dataset: params.dataset,
-        errorHeading: params.errorHeading,
-        issueItems: params.issueItems,
+        errorSummary: {
+          heading: params.errorSummary.heading,
+          items: params.errorSummary.items
+        },
```
[the remark this answers is not in the document]
Comment (GeorgeGoodall-GovUk) [reply test/unit/views/organisations/issueDetailsPage.test.js:182]: outside the scope of this ticket

File: src/middleware/dataview.middleware.js [hunk 3]
```diff
@@ -16,64 +14,37 @@ const validatedataviewQueryParams = validateQueryParams({
...
-    return next(new Error('Invalid page number'))
-  }
-
-  req.offset = (pageNumber - 1) * pageLength
-  next()
-}
-
 export const fetchEntities = fetchMany({
-  query: ({ req, params }) => `SELECT * FROM entity WHERE organisation_entity = ${req.orgInfo.entity} LIMIT ${pageLength} OFFSET ${req.offset}`,
+  query: ({ req, params }) => `SELECT * FROM entity WHERE organisation_entity = ${req.orgInfo.entity} LIMIT ${req.dataRange.pageLength} OFFSET ${req.dataRange.offset}`,
```
[the remark this answers is not in the document]
Comment (GeorgeGoodall-GovUk) [reply src/middleware/dataview.middleware.js:24]: should be fine as we getting from database

File: config/default.yaml [hunk 4]
```diff
@@ -88,3 +88,4 @@ guidanceNavigation:
       url: /guidance/publish-data-on-your-website
     - label: Keep your data up to date
       url: /guidance/keep-your-data-up-to-date
+tablePageLength: 50
```
Comment (DilwoarH) [inline config/default.yaml:91]: Do you need to update the schema on `config/utils.js`?

File: src/middleware/issueTable.middleware.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  req.templateParams = {
+    tableParams,
+    organisation: orgInfo,
+    dataset,
+    errorSummary,
+    issueType,
+    pagination,
+    dataRange
+    // issueEntitiesCount,
+    // pageNumber
```
Comment (DilwoarH) [inline src/middleware/issueTable.middleware.js:159]: won't these things break the page?

**Review by DilwoarH — CHANGES_REQUESTED** [review]

File: config/default.yaml (hunk 4 above)
Comment (GeorgeGoodall-GovUk) [reply config/default.yaml:91]: weird, I didn't get any valibot error??
I'll add it in now anyways

Comment (DilwoarH) [reply config/default.yaml:91]: yeah I raised this with @rosado a few times, I think we should make it throw an error

**Review by DilwoarH — APPROVED** [review]

**Pull request merged**